### PR TITLE
Vapor migration

### PR DESCRIPTION
**Pull request by josefdolezal**
Hello, Marcin 🖖!

Word has it the OnlinePlayground is still built on top of Kitura but is patiently waiting for someone to migrate it to Vapor. I've noticed a discussion on Twitter but didn't see any follow-up activity - so here I am, opening a midnight PR to make it happen.

---

I did my best keeping the PR small -- most of the changes just migrate the code ~1:1 to Vapor's ecosystem.

### Notable changes

* I migrated the project to Swift 5.7 and didn't see any benefit in keeping older versions around - so it's now the only supported version
* All static html pages now use templating engine - only assets (images, scripts, ...) are served directly from FS
* Everything related to building and running code snippets now lives in a dedicated package `BuildToolchainEngine`

### Missing parts

~~A fresh build on my M1 laptop works like a fine wine, but we are still few steps away from a production-ready build.~~

* ~~I am having hard time building a linux docker image - any help here is more than welcome~~
* ~~Tests are yet to be migrated~~
* Shutting down the server (^d) always ends with an error
    * This seems to be an [internal Vapor issue](https://github.com/vapor/vapor/issues/2502) when using web sockets - shutting down the server will always result in internal server error
    * Considering it only affects server shutdown, I wouldn't consider it to be a blocker

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,7 @@
 Toolchains/
 config/github.json
 *.xcodeproj
+.vscode/launch.json
```
Comment (josefdolezal) [inline .gitignore:7]: The `launch.json` was automatically generated by VSCode extension, so no need to keep it around

File: Sources/App/Middlewares/NotFoundMiddleware.swift [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import Vapor
+import Leaf
+
+public struct NotFoundMiddleware: AsyncMiddleware {
```
Comment (josefdolezal) [inline Sources/App/Middlewares/NotFoundMiddleware.swift:4]: Takes care of all unhandled requests by serving 404 static page.

File: Sources/App/Services/TerminalService.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
+import BuildToolchainEngine
+import TSCBasic
+import Vapor
+
+public final class TerminalService {
```
Comment (josefdolezal) [inline Sources/App/Services/TerminalService.swift:5]: Vapor does not (to my knowledge) have concept of "web-socket controller", so this class follows Kitura's API.

File: Sources/App/Services/TerminalService.swift [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            break
+        }
+    }
+
+    private func disconnected(on webSocket: WebSocket) {
+        close(webSocket: webSocket)
+    }
+
+    private func close(webSocket: WebSocket) {
+        connections.removeAll { $0 === webSocket }
```
Comment (josefdolezal) [inline Sources/App/Services/TerminalService.swift:72]: Not thread safe, potentially problematic

File: Sources/App/configure.swift [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import Vapor
+import Leaf
+
+public extension Application {
+    func configure() throws {
+        middleware.use(NotFoundMiddleware())
+        middleware.use(FileMiddleware(publicDirectory: directory.publicDirectory))
+        middleware.use(sessions.middleware)
```
Comment (josefdolezal) [inline Sources/App/configure.swift:8]: I've dropped the session's secret - is it ok?

File: Sources/App/routes.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+        get("privacy-policy") { req async throws -> View in
+            return try await req.view.render("privacy-policy")
+        }
+
+        // router.all
+        get("letsplay") { req -> Response in
+            req.session.data["playground"] = "true"
+            return req.redirect(to: "/")
+        }
```
Comment (josefdolezal) [inline Sources/App/routes.swift:21]: Serving `all` HTTP methods is not supported in Vapor, so I only implemented GET.

File: Sources/BuildToolchainEngine/BuildToolchain.swift [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            #endif
+            #if DEBUG
+                cmd += ["-I", projectDirectoryPath.appending(components: "OnlinePlayground", "OnlinePlayground-\(toolchain.rawValue)", ".build", "debug").pathString]
+                cmd += ["-L", projectDirectoryPath.appending(components: "OnlinePlayground", "OnlinePlayground-\(toolchain.rawValue)", ".build", "debug").pathString]
+            #else
+                cmd += ["-I", projectDirectoryPath.appending(components: "OnlinePlayground", "OnlinePlayground-\(toolchain.rawValue)", ".build", "release").pathString]
+                cmd += ["-L", projectDirectoryPath.appending(components: "OnlinePlayground", "OnlinePlayground-\(toolchain.rawValue)", ".build", "release").pathString]
+            #endif
+            cmd += ["-lOnlinePlayground"]
+//            cmd += ["-target", target]
```
Comment (josefdolezal) [inline Sources/BuildToolchainEngine/BuildToolchain.swift:83]: Target values hardcoded above failed the build on M1.

File: Sources/BuildToolchainEngine/BuildToolchain.swift [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            // cmd += ["-enforce-exclusivity=checked"] // needs -Onone
+            // Enable JSON-based output at some point.
+            // cmd += ["-parseable-output"]
+            if let sdkRoot = sdkRoot() {
+                cmd += ["-sdk", sdkRoot.pathString]
+            }
+            cmd += ["-o", binaryFilePath.pathString]
+            cmd += [mainFilePath.pathString]
+
+            print(cmd.joined(separator: " "))
```
Comment (josefdolezal) [inline Sources/BuildToolchainEngine/BuildToolchain.swift:104]: Forgotten comment 🙇‍♂️

File: Sources/BuildToolchainEngine/BuildToolchain.swift [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    // Path to toolchain, relative to current process PWD
+    var path: RelativePath {
+        RelativePath("Toolchains").appending(components: "swift-\(self.rawValue).xctoolchain", "usr", "bin")
+    }
+
+    var swift_version: String {
+        switch self {
+        // case .swift5_0:
+        //   return "5"
+        case .swift5_7:
```
Comment (josefdolezal) [inline Sources/BuildToolchainEngine/BuildToolchain.swift:27]: Removed old Swift versions (the toolchain is not available on silicon).

File: run.sh [hunk 10]
```diff
@@ -1,21 +1,14 @@
...
-    echo "Generate Xcode project..."
-    $SWIFT package generate-xcodeproj
-fi
-
 # Run
-$(npm bin)/webpack
+npx webpack
 
-$SWIFT run -c release --build-path .build/swift-$SWIFT_VERSION &
+$SWIFT run -c release --scratch-path .build/swift-$SWIFT_VERSION &
```
Comment (josefdolezal) [inline run.sh:10]: `--build-path` is deprecated

**Comment by darrellroot** [comment]
I'm not close to delivering anything useful (still experimenting/learning Vapor) so great job Josef!

**Comment by krzyzanowskim** [comment]
impressive work! keep up going. it's happening 🍭

File: Dockerfile [hunk 11]
```diff
@@ -47,4 +47,4 @@ RUN ./bootstrap.sh
 # CMD export PATH="$PATH:node_modules/.bin"
 # CMD export NVM_DIR="$HOME/.nvm"
 # CMD $NVM_DIR/nvm.sh
-CMD Toolchains/swift-5.7-RELEASE.xctoolchain/usr/bin/swift run -c release --static-swift-stdlib --build-path .build/swift-5.7-RELEASE
+CMD Toolchains/swift-5.7-RELEASE.xctoolchain/usr/bin/swift run -c release --scratch-path .build/swift-5.7-RELEASE PlaygroundServer serve --hostname 0.0.0.0
```
Comment (josefdolezal) [inline Dockerfile:50]: * Build fails when stdlib is linked statically (on aarch64), I am not sure about the consequences though
* Default hostname is 127.0.0.1, which is not accessible outside of the container

File: bootstrap.sh [hunk 12]
```diff
@@ -32,13 +32,16 @@ function install_toolchain {
             rm -r swift-$SWIFT_VERSION-$RELEASE-$SWIFT_TARGET.pkg
             ;;
         ubuntu*)
+            # select toolchain binary based on cpu architecture
+            if [[ $(arch) = aarch* ]]; then ARCH=-$(arch); fi
+
             mkdir -p Toolchains/swift-$SWIFT_VERSION-$RELEASE.xctoolchain
             # download
-            curl -LO https://download.swift.org/swift-$SWIFT_VERSION-$BRANCH/$LINUX_DISTRO/swift-$SWIFT_VERSION-$RELEASE/swift-$SWIFT_VERSION-$RELEASE-$SWIFT_TARGET.tar.gz
+            curl -LO https://download.swift.org/swift-$SWIFT_VERSION-$BRANCH/$LINUX_DISTRO$ARCH/swift-$SWIFT_VERSION-$RELEASE/swift-$SWIFT_VERSION-$RELEASE-$SWIFT_TARGET$ARCH.tar.gz
```
Comment (josefdolezal) [inline bootstrap.sh:40]: Universal binaries are not available on linux, so `-aarch64` is appended when building on arm

**Review by josefdolezal — COMMENTED** [review]
A lot of blood, sweat and tears went into it but I've managed to get the docker build work on M1. A test image is published under `josefdolezal/playground-vapor`, give it a spin 🤙.

**Comment by josefdolezal** [comment]
The whole app is now migrated and ready for review 👍

**Review by krzyzanowskim — APPROVED** [review]
This is fantastic!

**Comment by krzyzanowskim** [comment]
> I've noticed a discussion on Twitter but didn't see any follow-up activity - so here I am, opening a midnight PR to make it happen.

@josefdolezal In an [my tweet](https://twitter.com/krzyzanowskim/status/1568238041847562240) I offered the $$$, and I'd like to stand behind it. Let me know what/if you want to cash the reward!

**Comment by josefdolezal** [comment]
Don't worry about it! I had fun exploring the codebase 👍

**Pull request merged**